lib/Db/QuotaUsageMapper.php

```diff
@@ -0,0 +1,302 @@
+		$periodStart = (new DateTime())->sub(new DateInterval('P' . $timePeriod . 'D'))->getTimestamp();
+		
+		// Get the sum of the units used in the time period
+		$qb->select($qb->createFunction('SUM(units)'))
+			->from($this->getTableName())
+			->where(
+				$qb->expr()->eq('type', $qb->createNamedParameter($type, IQueryBuilder::PARAM_INT))
+			)
+			->andWhere(
+				$qb->expr()->eq('user_id', $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR))
+			)
+			->andWhere(
+				$qb->expr()->gt('timestamp', $qb->createNamedParameter($periodStart, IQueryBuilder::PARAM_INT))
+			);
+		
+		// Execute the query and return the result
+		$result = (int)$qb->executeQuery()->fetchOne();
+		$qb->resetQueryParts();
+		
+		return $result;
+	}
+
+	/**
+	 * @param string $userId
+	 * @param int $type
+	 * @return array
+	 * @throws Exception
+	 */
+	public function getQuotaUsagesOfUser(string $userId, int $type): array {
+		$qb = $this->db->getQueryBuilder();
+
+		$qb->select('*')
+			->from($this->getTableName())
+			->where(
+				$qb->expr()->eq('user_id', $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR))
+			)->andWhere(
+				$qb->expr()->eq('type', $qb->createNamedParameter($type, IQueryBuilder::PARAM_INT))
+			);
+		
+		$qb->orderBy('timestamp', 'DESC');
+
+		return $this->findEntities($qb);
+	}
+
+	/**
+	 * @param string $userId
+	 * @param int $type
+	 * @return int
+	 * @throws Exception
+	 * @throws \RuntimeException
+	 */
+	public function getQuotaUnitsOfUser(string $userId, int $type): int {
+		$qb = $this->db->getQueryBuilder();
+
+		$qb->select($qb->createFunction('SUM(units)'))
+			->from($this->getTableName())
+			->where(
+				$qb->expr()->eq('user_id', $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR))
+			)->andWhere(
+				$qb->expr()->eq('type', $qb->createNamedParameter($type, IQueryBuilder::PARAM_INT))
+			);
+
+		// Execute the query and return the result
+		$result = (int)$qb->executeQuery()->fetchOne();
+		$qb->resetQueryParts();
+		
+		return $result;
+	}
+
+	/**
+	 * @param string $userId
+	 * @param int $type
+	 * @param int $units
+	 * @return QuotaUsage
+	 * @throws Exception
+	 * @throws \RuntimeException
+	 * @throws DoesNotExistException
+	 * @throws MultipleObjectsReturnedException
+	 */
+	public function createQuotaUsage(string $userId, int $type, int $units): QuotaUsage {
+		$qb = $this->db->getQueryBuilder();
+
+		$timeStamp = (new DateTime())->getTimestamp();
+
+		$qb->insert($this->getTableName())
+			->values([
+				'type' => $qb->createNamedParameter($type, IQueryBuilder::PARAM_INT),
+				'user_id' => $qb->createNamedParameter($userId, IQueryBuilder::PARAM_STR),
+				'units' => $qb->createNamedParameter($units, IQueryBuilder::PARAM_INT),
+				'timestamp' => $qb->createNamedParameter($timeStamp, IQueryBuilder::PARAM_INT),
+			]);
+
+		$qb->executeStatement();
+		$qb->resetQueryParts();
+
+		return $this->getQuotaUsageOfUser((int)$qb->getLastInsertId(), $userId);
```

Comment:
This seems to duplicate QBMapper#insert(Entity)

Reply:
Hmm. True! This was my first dabble with having something stored in the db so I'll blame that.

Reply:
Thanks! Fixed in https://github.com/nextcloud/integration_openai/pull/54/commits/aa2052b27800e562f337992f530867d0f1e1b580